corehq/apps/case_importer/do_import.py

```diff
@@ -0,0 +1,278 @@
+
+from corehq.apps.export.tasks import add_inferred_export_properties
+from corehq.apps.hqcase.utils import submit_case_blocks
+from corehq.apps.locations.models import SQLLocation
+from corehq.apps.users.models import CouchUser
+from corehq.toggles import BULK_UPLOAD_DATE_OPENED
+from corehq.util.datadog.utils import case_load_counter
+from corehq.util.soft_assert import soft_assert
+
+from . import exceptions
+from . import util as importer_util
+from .const import LookupErrors
+
+CASEBLOCK_CHUNKSIZE = 100
+RowAndCase = namedtuple('RowAndCase', ['row', 'case'])
+
+
+class _Importer(object):
+    def __init__(self, domain, config, task, record_form_callback):
+        self.domain = domain
+        self.config = config
+        self.task = task
+        self._record_form_callback = record_form_callback
+
+        self.results = importer_util.ImportResults()
+
+        self.id_cache = {}
+        self.name_cache = {}
+        self.uncreated_external_ids = set()
+        self._caseblocks = []
+
+    @cached_property
+    def user(self):
+        return CouchUser.get_by_user_id(self.config.couch_user_id, self.domain)
+
+    def log_case_lookup(self):
+        case_load_counter("case_importer", self.domain)
+
+    def do_import(self, spreadsheet):
+        row_count = spreadsheet.max_row
+        for row_num, row in enumerate(spreadsheet.iter_row_dicts(), start=1):
+            set_task_progress(self.task, row_num, row_count)
+
+            # skip first row (header row)
+            if row_num == 1:
+                continue
+
+            try:
+                self.import_row(row_num, row)
+            except exceptions.CaseRowError as error:
+                self.results.add_error(row_num, error)
+
+        self.commit_caseblocks()
+
+    def import_row(self, row_num, raw_row):
+        search_id = importer_util.parse_search_id(self.config, raw_row)
+        fields_to_update = importer_util.populate_updated_fields(self.config, raw_row)
+        if not any(fields_to_update.values()):
+            # if the row was blank, just skip it, no errors
+            return
+
+        row = CaseImportRow(search_id, fields_to_update, self.config, self.domain, self.user.user_id)
+        row.check_valid_external_id()
+        if row.relies_on_uncreated_case(self.uncreated_external_ids):
+            self.commit_caseblocks()
+
+        case, error = importer_util.lookup_case(
+            self.config.search_field,
+            row.search_id,
+            self.domain,
+            self.config.case_type
+        )
+        self.log_case_lookup()
+
+        if error == LookupErrors.NotFound:
+            if not self.config.create_new_cases:
+                return
+        elif error == LookupErrors.MultipleResults:
+            raise exceptions.TooManyMatches()
+
+        row.set_owner_id(self.name_cache, self.id_cache)
+        row.set_parent_id(self.log_case_lookup)
+        row.set_date_opened()
+        row.set_external_id(is_new_case=not case)
```

Comment:
I wrote a comment about this exact block, actually, but forgot to submit it.  This is one of the remaining pieces that stands out as ugly to me too.  What I'd really like to do with this is pass in everything `CaseImportRow` needs up-front, then thereafter only access getter methods, without changing state internal to `CaseImportRow`.  Unfortunately that'd require passing in like 12 parameters to the constructor, which is also gross.  

For now I kept control flow here, where all the necessary objects are.  Can you think of a good way to move control flow to the other class without giving it access to essentially all the attributes of this parent class?

One possible approach I considered would be to group all these dependencies,  like `self.results` - say, a object that handles caching, and a light wrapper around metadata stuff like domain, user, and config.  That doesn't really reduce the coupling between the classes, but it would compartmentalize it a bit.

Reply:
Just saw your other comment:
> building on my comment from before - I think there might be a way that you can just generate these when you call self.extras, so it would be self.get_extras() which would then return all of the appropriate items so that the _Importer class doesn't have to populate them.

I definitely agree.  Do you think it'd be preferable to pass in everything up front and avoid mutating the class?  That's kind of where I'm leaning now, though I would like to somehow minimize the dependencies between the two classes.

Reply:
I'd second the approach to avoid mutating the class. I'd argue that it makes it more testable and easier to reason about, even if passing stuff in up-front could be ugly.

Reply:
I think both of the caches can be extracted out to use quickcache so you don't have to pass the dicts around. 
In general though I also think it's preferable to avoid mutating the class from outside.

Reply:
> I think both of the caches can be extracted out to use quickcache so you don't have to pass the dicts around.

I'm not sure - right now it's an in-memory lookup, and switching to quickcache would mean using redis or memcached.  Tying the cache to this code means it can persist for as short or as long as the case import takes (which can vary considerably), and the lookups are always in memory.

I also considered making a object to store these caches and the associated code, say `CaseOwnerAccessor` or something.  I do think that'd be a good way to further simplify this code, although I'm gonna hold off on that, as I'm trying to somewhat limit the scope of this refactor.